src/v/cloud_storage/partition_manifest.cc

```diff
@@ -932,121 +932,163 @@ bool partition_manifest::add(
     return add(m);
 }
 
-bool partition_manifest::safe_segment_meta_to_add(const segment_meta& m) {
-    if (_segments.empty()) {
-        // The empty manifest can be started from any offset. If we deleted all
-        // segments due to retention we should start from last uploaded offset.
-        // The reuploads are not possible if the manifest is empty.
-        const bool is_safe = _last_offset == model::offset{0}
-                             || model::next_offset(_last_offset)
-                                  == m.base_offset;
-
-        if (!is_safe) {
-            vlog(
-              cst_log.error,
-              "[{}] New segment does not line up with last offset of empty "
-              "log: "
-              "last_offset: {}, new_segment: {}",
-              _ntp,
-              _last_offset,
-              m);
-        }
-
-        return is_safe;
-    }
-
-    auto format_seg_meta_anomalies = [](const segment_meta_anomalies& smas) {
-        if (smas.empty()) {
-            return ss::sstring{};
-        }
-
-        std::vector<anomaly_type> types;
-        for (const auto& a : smas) {
-            types.push_back(a.type);
-        }
+size_t partition_manifest::safe_segment_meta_to_add(
+  std::vector<segment_meta> meta_list) const {
+    struct manifest_substitute {
+        model::offset last_offset;
+        std::optional<segment_meta> last_segment;
+        size_t num_accepted{0};
+    };
 
-        return ssx::sformat(
-          "{{anomaly_types: {}, new_segment: {}, previous_segment: {}}}",
-          types,
-          smas.begin()->at,
-          smas.begin()->previous);
+    manifest_substitute subst{
+      .last_offset = _last_offset,
+      .last_segment = last_segment(),
     };
 
-    auto it = _segments.find(m.base_offset);
-    if (it == _segments.end()) {
-        // Segment added to tip of the log
-        const auto last_seg = last_segment();
-        vassert(last_seg.has_value(), "Empty manifest");
+    for (const auto& m : meta_list) {
+        if (_segments.empty() && !subst.last_segment.has_value()) {
```

Comment:
nit: just curious, did we actually need both of these conditions here? Would just `!subst.last_segment.has_value()` have sufficed?